File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/prototype/lab/garden.cpp
// garden.cpp

#include "conventional_lettuce.h"
#include "dp_exception.h"
#include "lettuce.h"
#include "organic_lettuce.h"
#include "garden.h"

garden::garden(lettuce* a_lettuce_prototype)
{
	set_lettuce_prototype(a_lettuce_prototype);
}

void garden::set_lettuce_prototype(lettuce* a_lettuce_prototype)
{
	if (!a_lettuce_prototype) 
	{
            throw dp_exception("a_lettuce_prototype is NULL!");
	}
	my_lettuce_prototype = a_lettuce_prototype->clone();
}

lettuce * garden::make_lettuce() const
{
	return my_lettuce_prototype;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/prototype/solution/Garden.java
// Garden.java

public class Garden
{
	public Garden(Lettuce aLettucePrototype)
	{
		setLettucePrototype(aLettucePrototype);
	}

	public void setLettucePrototype(Lettuce aLettucePrototype)
	{
		if (aLettucePrototype == null)
		{
			throw new DpException("void setLettucePrototype(Lettuce aLettucePrototype) -- aLettucePrototype == null.");
		}

		myLettucePrototype = aLettucePrototype.clone();
	}

	public Lettuce makeLettuce()
	{
		return myLettucePrototype.clone();
	}

	private Lettuce myLettucePrototype;
}
